File: Sources/TimeoutIntegrations/AppRoot.swift
import AppKit
import TimeoutEngine

private let bundleId = "com.aurelius.timeout"

/// 应用根装配点：装配引擎 + 心跳 + 系统采样 + 三大 controller，并注册睡眠/唤醒观察者。
/// 启动加载持久化 config/state；运行期节流持久化状态；退出时落盘。
final public class AppRoot {
    public static let shared = AppRoot()

    private var statusItem: StatusItemController?
    private var engine: LiveTimeoutEngine?
    private var overlayController: LiveOverlayController?
    private var calendarProvider: LiveCalendarProvider?
    private var heartbeat: HeartbeatTimer?
    private var sensors: SystemSensors?
    private var configStore: ConfigStore?
    private var settingsController: SettingsWindowController?
    private var sleepObservers: [NSObjectProtocol] = []

    private var lastSavedPhase: EnginePhase?
    private var lastSaveAt: Date = .distantPast

    private init() {}

    public func start() {
        AccessibilityChecker.bootstrap()  // 引导 Accessibility 授权（媒体键控制必需）

        let dir = ConfigStore.defaultDirectory(bundleId: bundleId)
        let store: ConfigStore?
        do {
            store = try ConfigStore(directory: dir)
        } catch {
            NSLog("[Timeout] 持久化目录初始化失败，本次运行不落盘：\(error)")
            store = nil
        }
        configStore = store

        let config = debugConfigOrLoaded(store: store)
        let sensors = SystemSensors()
        self.sensors = sensors

        let overlay = LiveOverlayController()
        overlay.onRequestEarlyExit = { [weak self] in self?.engine?.requestEarlyRestExit() }
        self.overlayController = overlay

        let calendar = LiveCalendarProvider()
        calendar.bootstrap()  // 触发日历权限请求（用户手动授予）
        self.calendarProvider = calendar

        let engine = LiveTimeoutEngine(
            clock: SystemClock(),
            calendarProvider: calendar,
            overlay: overlay,
            music: LiveMusicController(),
            systemState: sensors,
            config: config,
            initialState: store?.loadState()
        )
        engine.fastForward()  // 启动崩溃恢复
        engine.setPersistHandler { [weak self] state in self?.handlePersist(state) }
        self.engine = engine
        lastSavedPhase = engine.state.phase

        statusItem = StatusItemController(
            onForceRest: { [weak self] in
                self?.engine?.forceRestNow()
                self?.engine?.tick()  // 立即生效，不等下一秒心跳
            },
            loginEnabled: LoginService.isEnabled,
            onSetLaunchAtLogin: { LoginService.setEnabled($0) },
            onOpenSettings: { [weak self] in self?.openSettings() }
        )

        let heartbeat = HeartbeatTimer(queue: .main)  // 主队列：副作用（overlay/music）均 UI 安全
        heartbeat.start(interval: 1.0) { [weak self] in
            guard let self else { return }
            self.engine?.tick()
            if let phase = self.engine?.state.phase {
                self.statusItem?.setStatus(text: self.statusText(for: phase, engine: self.engine))
            }
        }
        self.heartbeat = heartbeat

        settingsController = SettingsWindowController { [weak self] newConfig in
            guard let self else { return }
            if let store = self.configStore {
                do { try store.saveConfig(newConfig) }
                catch { NSLog("[Timeout] 配置保存失败：\(error.localizedDescription)") }
            }
            self.engine?.updateConfig(newConfig)
            NSLog("[Timeout] 配置已应用：\(newConfig.workWindows.count) 个工作窗口 / 工作 \(Int(newConfig.workIntervalSeconds/60))min / 休息 \(Int(newConfig.restDurationSeconds/60))min")
        }

        registerSleepObservers()
        NSLog("[Timeout] 引擎启动 phase=\(engine.state.phase.rawValue) accum=\(Int(engine.state.workAccumulatedSeconds))s")

        // 调试：启动即打开设置窗（便于截图验证 UI）
        if ProcessInfo.processInfo.environment["TIMEOUT_SHOW_SETTINGS"] != nil {
            openSettings()
        }
    }

    /// 应用退出前落盘最终状态。
    public func shutdown() {
        if let state = engine?.state { configStore?.saveState(state) }
        heartbeat?.stop()
        for observer in sleepObservers {
            NSWorkspace.shared.notificationCenter.removeObserver(observer)
        }
        sleepObservers.removeAll()
    }

    // MARK: - 设置窗口

    func openSettings() {
        guard let config = engine?.config else { return }
        settingsController?.show(currentConfig: config)
    }

    // MARK: - 调试配置

    /// TIMEOUT_DEBUG=1 时使用极速配置（8s 工作 / 15s 休息 / 全天窗口 / 禁用 AFK）便于手动验证遮罩与音乐。
    private func debugConfigOrLoaded(store: ConfigStore?) -> DayPlanConfig {
        if ProcessInfo.processInfo.environment["TIMEOUT_DEBUG"] != nil {
            var c = DayPlanConfig.defaultConfig
            c.workIntervalSeconds = 8
            c.restDurationSeconds = 15
            c.afkThresholdSeconds = 999_999
            c.workWindows = [WorkWindow(start: TimeOfDay(hours: 0), end: TimeOfDay(hours: 23, minutes: 59))]
            NSLog("[Timeout] DEBUG 模式：8s 工作 / 15s 休息 / 全天窗口")
            return c
        }
        return store?.loadConfig() ?? .defaultConfig
    }

    // MARK: - 节流持久化（phase 变化或每 5s 落盘一次）

    private func handlePersist(_ state: EngineState) {
        let now = Date()
        let phaseChanged = state.phase != lastSavedPhase
        guard phaseChanged || now.timeIntervalSince(lastSaveAt) > 5 else { return }
        configStore?.saveState(state)
        lastSavedPhase = state.phase
        lastSaveAt = now
    }

    // MARK: - 菜单栏倒计时文案

    private func statusText(for phase: EnginePhase, engine: LiveTimeoutEngine?) -> String {
        guard let engine else { return "🍅" }
        switch phase {
        case .working:
            let remain = max(0, engine.config.workIntervalSeconds - engine.state.workAccumulatedSeconds)
            return "工作 \(Int(ceil(remain / 60)))′"
        case .resting:
            guard let start = engine.state.restStartedAt else { return "休息中" }
            let deadline = start.addingTimeInterval(engine.config.restDurationSeconds)
            let remain = max(0, deadline.timeIntervalSince(Date()))
            return "休息 \(Int(ceil(remain / 60)))′"
        case .inMeeting: return "会议中"
        case .idle: return "暂停"
        case .offDuty: return "下班"
        }
    }

    // MARK: - 睡眠 / 唤醒

    private func registerSleepObservers() {
        let nc = NSWorkspace.shared.notificationCenter
        let will = nc.addObserver(forName: NSWorkspace.willSleepNotification, object: nil, queue: .main) { [weak self] _ in
            self?.sensors?.isAsleep = true
            self?.engine?.handleSleep()
            self?.heartbeat?.suspend()
            NSLog("[Timeout] 系统睡眠：挂起心跳")
        }
        let did = nc.addObserver(forName: NSWorkspace.didWakeNotification, object: nil, queue: .main) { [weak self] _ in
            self?.sensors?.isAsleep = false
            self?.heartbeat?.resume()
            self?.engine?.handleWake()
            NSLog("[Timeout] 系统唤醒：恢复心跳 + 重置对账基点")
        }
        sleepObservers = [will, did]
    }
}

File: windows/TimeoutShell/AppRoot.cs
using System.Windows.Threading;
using TimeoutEngine;
using TimeoutEngine.Win32;
using TimeoutShell.Adapters;
using TimeoutShell.Power;
using TimeoutShell.Tray;
using TimeoutEngine.Graph;

namespace TimeoutShell;

// 装配点 · 镜像 Swift AppRoot.swift。汇聚 6 个 interface 实现 → 注入 LiveTimeoutEngine →
// 崩溃恢复 → 持久化 handler → 电源事件 → 心跳 1Hz tick → 托盘（非 headless）。
public sealed class AppRoot : IDisposable
{
    private readonly ConfigStore _store;
    private DayPlanConfig _config;
    private readonly WindowsSystemState _systemState = new();
    private readonly ICalendarProvider _calendar;
    private readonly FullscreenOverlayController _overlay;
    private readonly WasapiAmbientPlayer _ambient;
    private readonly MusicController _music;
    private readonly HeartbeatTimer _heartbeat;
    private LiveTimeoutEngine _engine = null!;
    private PowerEventBridge? _power;
    private TrayController? _tray;

    public AppRoot(Dispatcher dispatcher)
    {
        string dir = ConfigStore.DefaultDirectory("com.aurelius.timeout");
        _store = new ConfigStore(dir);
        _config = _store.LoadConfig();
        if (Environment.GetEnvironmentVariable("TIMEOUT_DEBUG") == "1") ApplyDebugConfig();

        _calendar = BuildCalendarProvider(_config);
        _ambient = new WasapiAmbientPlayer(new PinkNoiseGenerator());
        _music = new MusicController(_ambient,
            new MediaKeySender(new PInvokeSendInputPort()),
            new QqMusicDetector(new ProcessNameProvider()));
        _overlay = new FullscreenOverlayController(dispatcher);
        _heartbeat = new HeartbeatTimer(dispatcher);
    }

    public LiveTimeoutEngine Engine => _engine;

    public void Start(bool headless)
    {
        _engine = new LiveTimeoutEngine(
            clock: new SystemClock(),
            calendarProvider: _calendar,
            overlay: _overlay,
            music: _music,
            systemState: _systemState,
            config: _config,
            initialState: _store.LoadState());
        _overlay.OnRequestEarlyExit = () => _engine.RequestEarlyRestExit();  // Esc 双击 → 引擎内部 Dismiss（不自 Dismiss）
        _engine.SetPersistHandler(s => _store.SaveState(s));
        _engine.FastForward();

        _power = new PowerEventBridge(_engine, _heartbeat, _systemState);
        _power.Start();

        if (!headless)
            _tray = new TrayController(_engine, () => Console.WriteLine("[Timeout][settings] (Phase 1 占位)"));

        _heartbeat.Start(1.0, () => _engine.Tick());
        Console.WriteLine($"[Timeout][root] ASSEMBLY_OK 装配完成 headless={headless} interval={_config.WorkIntervalSeconds}s rest={_config.RestDurationSeconds}s");
    }

    /// <summary>TIMEOUT_DEBUG 极速配置：使 CI 烟测在 5s 内见证 working→resting→working 相位转移。</summary>
    private void ApplyDebugConfig()
    {
        _config = new DayPlanConfig
        {
            WorkWindows = new() { new WorkWindow(new TimeOfDay(0, 0), new TimeOfDay(23, 59)) },
            WorkIntervalSeconds = 2,        // 2s 工作
            RestDurationSeconds = 1,        // 1s 休息
            AfkThresholdSeconds = 999_999,  // 抑制 idle 干扰（runner 无人操作），确保触发相位转移
            AmbientSoundEnabled = false,    // headless 无音频设备，避免 WASAPI 异常
            ControlQQMusic = false,
        };
        Console.WriteLine("[Timeout][debug] TIMEOUT_DEBUG 极速配置（2s 工作 / 1s 休息）");
    }

    /// <summary>按 graphClientId 条件注入日历门控；空/失败 → EmptyCalendarProvider 降级（headless 不崩）。</summary>
    private ICalendarProvider BuildCalendarProvider(DayPlanConfig config)
    {
        var clientId = config.GraphClientId;
        if (string.IsNullOrWhiteSpace(clientId))
        {
            Console.WriteLine("CALENDAR_DEGRADED graphClientId 为空，日历门控降级（无会议）");
            return new EmptyCalendarProvider();
        }
        try
        {
            var client = new MsalGraphClient(clientId);
            if (!client.IsConfigured) return new EmptyCalendarProvider();
            Console.WriteLine($"[Timeout][calendar] Graph 日历门控已装配（client id 末 4 位：…{clientId[^4..]}）");
            return new GraphCalendarProvider(client, new SystemClock());
        }
        catch (Exception ex)
        {
            Console.WriteLine($"CALENDAR_DEGRADED Graph 装配失败，降级：{ex.Message}");
            return new EmptyCalendarProvider();
        }
    }

    public void Dispose()
    {
        _heartbeat.Stop();
        _power?.Dispose();
        _tray?.Dispose();
        _overlay.Dispose();   // 卸键盘钩子 + 关遮罩窗
        _music.Dispose();
    }
}
